README.md
# qisjob

## Qiskit Job Control

* [The `qisjob` module](#the-qisjob-module)
* [The `qisjob` command](#the-qisjob-command)
* [The `QisJob` class](#the-qisjob-class)
* [`QisJob` Documentation](#qisjob-documentation)
* [Prerequisites](#prerequisites)
* [Install](#install)
* [Usage](#usage)
* [Notes](#notes)

## The `qisjob` module

The module is `qisjob`. It was formerly named `qis_job`. I'm not sure why, but I have now changed it to `qisjob`.

If you previously installed the module as `qis_job` you can uninstall that older version either by

* `pip3 uninstall qis_job`
* `make uninstall_oldname`

## The `qisjob` command

The `qisjob` command loads and executes [Qiskit](https://qiskit.org) experiments on simulators or on genuine quantum
computing hardware such as that found at [IBM Q Experience](https://quantum-computing.ibm.com). Input is from one or
more OpenQASM 2 (by default) or OpenQASM 3 (with the `--qasm3_in` option) source files or Qiskit `QuantumCircuit`
Python code (with the `--qc` option) or from source provided via standard input in the absence of file arguments.

The command also provides some utility functions such as:

* enumerating backend platforms
* configuration both current and historical of backend platforms
* status of backend platforms
* status and results of jobs both current and historical

and other useful operations for Qiskit experimentation.

**`qisjob` comes with NO GUARANTEE and NO WARRANTY including as regards correctness nor applicability.
See [LICENSE](https://github.com/jwoehr/qisjob/blob/main/LICENSE.txt).**

`qisjob` can run Qiskit experiments expressed as either:

* [OPENQASM Open Quantum Assembly Language](https://github.com/openqasm/openqasm)
  * Use a well-formed OPENQASM file.
  * Examples are found in the `qasm_examples` [subdirectory](https://github.com/jwoehr/qisjob/tree/master/share/qasm_examples) of the project.
  * Examples of OpenQASM 3 are found in the `qasm3_examples` [subdirectory](https://github.com/jwoehr/qisjob/tree/qasm3/share/qasm3_examples)
  of the project, as well as in the   [OpenQASM project examples directory](https://github.com/openqasm/openqasm/tree/main/examples).
* Qiskit Terra `QuantumCircuit` Python code snippet.
  * To use a code snippet, only import that which is absolutely needed in the snippet and provide no execution code.
  * Pass the name of your `QuantumCircuit` to the `--qc` argument of `qisjob.py`
    * If you have multiple files of this sort, all must have the same name for their `QuantumCircuit` object.
    * An example circuit (very long execution!) is found in the `qc_examples` [subdirectory](https://github.com/jwoehr/qisjob/tree/master/share/qc_examples) of the project.

You can load and run multiple files, but you cannot mix Qasm and `QuantumCircuit` files in the same execution of the `qisjob`.

`QisJob` is compatible with the experimental [NuQasm2](https://github.com/jwoehr/nuqasm2) project, that you can use to compile and run your OPENQASM2.0 source code.
Given that you have `NuQasm2` installed, you can use `qisjob`'s `-n` _include-path:include-path:..._ switch

## The `QisJob` class

The `qisjob` script works by instancing an object of the class `qisjob.QisJob`.

You can use an object instance of the class `qisjob.QisJob` in your own program for its utility functions or the execute
experiments on real quantum hardware and/or simulators either using OpenQASM source or Qiskit `QuantumCircuit` source code.

### `QisJob` Documentation

The `qisjob.QisJob` object that underlies the `qisjob` script that can be used in your own programs has many args/kwargs.

The documentation for `qisjob.QisJob` is installed with the module in the site library's `share/doc`.

You can also read the [QisJob Documentation Online](http://www.softwoehr.com/softwoehr/oss/qisjob/)

## Prerequisites

* [Qiskit/qiskit-terra](https://github.com/Qiskit/qiskit-terra)
* [Qiskit/qiskit-aer](https://github.com/Qiskit/qiskit-aer)
* A provider such as [Qiskit/qiskit-ibmq-provider](https://github.com/Qiskit/qiskit-ibmq-provider)
* Currently supported backend providers are:
  * IBMQ (required)
    * For the local Aer simulator you only need qiskit-aer installed.
      * For genuine QPU or cloud simulator you will need an [IBM Q Experience API token](https://quantum-computing.ibm.com/account).
  * Forest (optional; a warning message may appear if absent)
    * You need [quantastic/qiskit-forest](https://github.com/quantastica/qiskit-forest)
      * `pip install quantastica-qiskit-forest`
    * For Rigetti QPU you will need [access](https://qcs.rigetti.com/request-access)
  * MQT (optional; a warning message may appear if absent)
    * You need [cda-tum/ddsim](https://github.com/cda-tum/ddsim)
    * This was formerly the JKU simulator
  * QI (optional; a warning message may appear if absent)
    * Install QuTech-Delft/quantuminspire from either
      * [Github QuTech-Delft/quantuminspire](https://github.com/QuTech-Delft/quantuminspire)
        * `pip install quantuminspire`.
      * You will also need a [Quantum Inspire token](https://www.quantum-inspire.com/account).

## Install

Do one of the following in the source directory (preferably in a Python virtual environment set up for Qiskit)

* `make # gnu make, we have provided a Makefile`
* `./setup.py install`
* `pip3 install .`

**Note** that the module name has recently (2021-02-16) changed from `qis_job` to `qisjob`. If you previously installed the module as `qis_job` you can uninstall that older version either by

* `pip3 uninstall qis_job`
* `make uninstall_oldname`

## Uninstall

Do one of the following in the source directory

* `make uninstall`
* `pip3 uninstall qisjob`

## Usage

The `qisjob` script has helpful help.

```text
$ qisjob --help
usage: qisjob [-h] [-i | -s | -a | -b BACKEND]
              [--qasm_simulator | --unitary_simulator | --statevector_simulator | --pulse_simulator | --densitymatrix_simulator]
              [--statevector_gpu | --unitary_gpu | --density_matrix_gpu] [--fakenoise FAKENOISE] [--display] [--version]
              [--api_provider API_PROVIDER] [--hub HUB] [--group GROUP] [--project PROJECT] [--providers] [--qvm] [--qvm_as] [--backends] [-1]
              [-d DATETIME] [-g] [-j] [--jobs JOBS] [--job_id JOB_ID] [--job_result JOB_RESULT] [-m] [-n NUQASM2] [-o OUTFILE] [-p] [-q QUBITS]
              [--qiskit_version] [-r] [-t SHOTS] [-v] [-x] [--showsched] [--circuit_layout] [--optimization_level OPTIMIZATION_LEVEL]
              [--histogram] [--plot_state_city PLOT_STATE_CITY] [--figure_basename FIGURE_BASENAME] [--qasm] [--qc QC] [--status]
              [--token TOKEN] [--url URL] [--use_job_monitor] [--job_monitor_line JOB_MONITOR_LINE]
              [--job_monitor_filepath JOB_MONITOR_FILEPATH] [-w] [--qasm3_in] [--qasm3_out]
              [filepath ...]

Qisjob loads from one or more OpenQASM source files or from a file containing a Qiskit QuantumCircuit definition in Python and runs as
experiments with reporting in CSV form. Can graph results as histogram or state-city plot. Also can give info on backend properties, qiskit
version, show circuit transpilation, etc. Can run as multiple jobs or all as one job. Exits 0 on success, 1 on argument error, 100 on runtime
error, 200 on QiskitError. Copyright 2019, 2023 Jack Woehr jwoehr@softwoehr.com PO Box 82, Beulah, Colorado 81023. Apache License, Version 2.0
-- See LICENSE which you should have received with this code. Unless required by applicable law or agreed to in writing, software distributed
under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License
for the specific language governing permissions and limitations under the License.

positional arguments:
  filepath              Filepath(s) to 0 or more .qasm files, default is stdin

options:
  -h, --help            show this help message and exit
  -i, --ibmq            Use best genuine IBMQ processor (default)
  -s, --sim             Use IBMQ qasm simulator
  -a, --aer             Use QISKit Aer simulator. Default is Aer simulator. Use -a --qasm-simulator to get Aer qasm simulator. Use -a --unitary-
                        simulator to get Aer unitary simulator.
  -b BACKEND, --backend BACKEND
                        Use specified IBMQ backend
  --qasm_simulator      With -a use qasm_simulator instead of aer simulator
  --unitary_simulator   With -a use unitary_simulator instead of aer simulator
  --statevector_simulator
                        With -a use aer_simulator_statevector instead of aer simulator
  --pulse_simulator     With -a use pulse_simulator instead of aer simulator
  --densitymatrix_simulator
                        With -a use aer_simulator_density_matrix instead of aer simulator
  --statevector_gpu     With -a and --qasm_simulator use gpu statevector simulator
  --unitary_gpu         With -a and --qasm_simulator use gpu unitary simulator
  --density_matrix_gpu  With -a and --qasm_simulator use gpu density matrix simulator
  --fakenoise FAKENOISE
                        Uses IBM/fake backend to simulate noise in the circuit with -a use --fakenoise ibmq_lima/ --fakenoise FakeVigo
  --display             Uses circuit.draw to visualize the untranspiled circuit
  --version             Announce QisJob version
  --api_provider API_PROVIDER
                        Backend remote api provider, currently supported are [IBMQ | Forest | MQT | QI]. Default is IBMQ.
  --hub HUB             Provider hub, default is 'ibm-q'
  --group GROUP         Provider group, default is 'open'
  --project PROJECT     Provider project, default is 'main'
  --providers           List hub/group/project providers for IBMQ
  --qvm                 Use Forest local qvm simulator described by -b backend, generally one of qasm_simulator or statevector_simulator. Use
                        --qvm_as to instruct the simulator to emulate a specific Rigetti QPU
  --qvm_as              Use Forest local qvm simulator to emulate the specific Rigetti QPU described by -b backend. Use --qvm to run the Forest
                        local qvm simulator described by -b backend.
  --backends            Print list of backends to stdout and exit 0
  -1, --one_job         Run all experiments as one job
  -d DATETIME, --datetime DATETIME
                        Datetime 'year,month,day[,hour,min,sec]' for -p,--properties
  -g, --configuration   Print configuration for backend specified by -b to stdout and exit 0
  -j, --job             Print your job's dictionary
  --jobs JOBS           Print JOBS number of jobs and status for -b backend and exit 0
  --job_id JOB_ID       Print job number JOB_ID for -b backend and exit 0
  --job_result JOB_RESULT
                        "Print result of job number JOB_RESULT for -b backend and exit 0
  -m, --memory          Print individual results of multishot experiment
  -n NUQASM2, --nuqasm2 NUQASM2
                        "Use nuqasm2 to translate OPENQASM2 source, providing include path for any include directives
  -o OUTFILE, --outfile OUTFILE
                        Write appending CSV to outfile, default is stdout
  -p, --properties      Print properties for backend specified by -b to stdout and exit 0
  -q QUBITS, --qubits QUBITS
                        Number of qubits for the experiment, default is 5
  --qiskit_version      Print Qiskit version and exit 0
  -r, --result          Print job result
  -t SHOTS, --shots SHOTS
                        Number of shots for the experiment, default 1024, max 8192
  -v, --verbose         Increase runtime verbosity each -v up to 3. If precisely 4, prettyprint QisJob's data dictionary and return (good for
                        debugging script arguments)
  -x, --transpile       Print circuit transpiled for chosen backend to stdout before running job
  --showsched           In conjuction with -x, show schedule for transpiled circuit for chosen backend to stdout before running job
  --circuit_layout      With -x, write image file of circuit layout after transpile (see --figure_basename)
  --optimization_level OPTIMIZATION_LEVEL
                        Set optimization level for transpilation before run, valid values 0-3, default is 1
  --histogram           Write image file of histogram of experiment results (see --figure_basename)
  --plot_state_city PLOT_STATE_CITY
                        Write image file of state city plot of statevector to PLOT_STATE_CITY decimal points (see --figure_basename)
  --figure_basename FIGURE_BASENAME
                        basename including path (if any) for figure output, default='figout', backend name, figure type, and timestamp will be
                        appended
  --qasm                Print qasm file to stdout before running job
  --qc QC               Indicate variable name of Python-coded QuantumCircuit
  --status              Print status of chosen --backend to stdout (default all backends) of --api_provider (default IBMQ) and exit 0
  --token TOKEN         Use this token
  --url URL             Use this url
  --use_job_monitor     Display job monitor instead of just waiting for job result
  --job_monitor_line JOB_MONITOR_LINE
                        Comma-separated list of hex values for character(s) to emit at the head of each line of job monitor output, default is
                        '0x0d'
  --job_monitor_filepath JOB_MONITOR_FILEPATH
                        Filepath for Job Monitor output if Job Monitor requested by --use_job_monitor, default is sys.stdout
  -w, --warnings        Don't print warnings on missing optional modules
  --qasm3_in            Interpret input as OpenQASM 3
  --qasm3_out           Print qasm file as OpenQASM 3 to stdout before running job
```

## Notes

* It is recommended you download or clone the most recent [release](https://github.com/jwoehr/qisjob/releases).
  * Version v4.1.2 is the current version.
* This document always reflects the latest checkins and may be ahead of the release versions.
  * See the README.md in the release version itself for contemporary information.
* Please use the [issue tracker](https://github.com/jwoehr/qisjob/issues) to report any issues or feature requests.
* If Python complains about the certs, you could try setting an env variable, like this:
  * `export SSL_CERT_FILE=$(python3 -c "import certifi; print(certifi.where())")`

## Software Bill of Materials

The [Software Bill of Materials (SBOM)](https://docs.github.com/en/code-security/supply-chain-security/understanding-your-software-supply-chain/exporting-a-software-bill-of-materials-for-your-repository) is `SBOM_qisjob_jwoehr_*.json`

Jack Woehr 2023-06-07


CHANGES.txt
QisJob https://github.com/jwoehr/qisjob

Apache License, Version 2.0 -- See LICENSE which you should have received with this code.

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

CHANGES by revision to QisJob

v4.1.3-dev
- Remove support for deprecated provider
  - For IBM Quantum only support `qiskit-ibm-provider`
- Import `QiskitError` from `qiskit.exception` instead of from `qiskit`
- Suppress warnings now working/ Imports location changed to only when needed
- Added --display to display the quantum circuit
- Updated and moved Qiskit Aer to qisjobaer.py 
- Moved Noise simulator to qisjobaer.py
- Ceated a test_aer.py file for testing of aer.

v4.1.2
- Use IBMProvider if installed
- Add OPENQasm 3 support
- Added `entangle.qasm` example
- Update alternate providers (e.g., JKU -> MQT)
- General code cleanups

v4.1.1
This is an intermediate release as the Qiskit BaseBackend, BaseJob, and
BaseProvider classes have all been removed in favor of BackendV2, JobV1, and
ProviderV1. The max_credits option has also been deprecated in Qiskit and is
removed from QisJob

v4.1
Note: Version 4.1 is the last version before changing Aer simulator support to the new `AerSimulator` backend.
- Add support for IBMQ hub, group, and project (`--hub` `--group` `--project`)
- Add support for `job_monitor` kwarg `line_discipline` (`--job_monitor_line`)
- Add `--noisy_sim` for noisy simulations
- Add `--providers` to list hub/group/project providers for IBMQ
- 4 `-v`s now print QisJob data dictionary and return
- `ibmqjob_to_dict` calls on members not supported on all platforms: Now each
  call is wrapped in try/except.
- `examples/multimeasure.qasm`
- Correct version number in `setup.py`

v4.0
- Renamed module to `qisjob` from `qis_job` (Issue #3)
	- Added Makefile target `uninstall_oldname` to uninstall `qis_job`
- `qisjob.QisJob` is now fully usable class to include in your Python program.
	- Added exception regimen and removee `sys.exit()` calls (Issue #17)
	Thanks, Omar Costa Hamido! (github `omarcostahamido`)
- Add `--job_monitor_filepath`
- Use `Aer` instead of `BasicAer` (Issue #15)
- Fix bug in `-w` switch (Issue #6)
- Moved `ibmqjob_to_dict()` into `QisJob` as static method
- Better install instructions (Issue #5)
- Autogenerate docs using `pdoc3`

v3.3:
	- `--optimization_level` argument
	- QI API changed, qisjob now automatically picks a QI backend if none named
	- `BackendStatus` changed in Qiskit so changed `--status` code
	- Save result_exp as QisJob.result_exp
	- If -j then print job both before and after run
	- More verbosity info on Aer jobs (gpu)
	- Add `yiqing_5.qasm` to test QuantumInspire Starmon-5
	- Change `--qisjob_version` to `--version`
	- Change "master" branch to "main"
	- Update readme to make it clearer that qisjob can take input from stdin
	- `-w, --warnings` suppress warnings of missing optional modules

v3.2:
	- QisJob can be instanced with qasm source using `qasm_src=` a string
	- More support for Aer gpu
	- Job monitor no longer default, if desired, `--use_job_monitor`

v3.1:
	- Created Makefile
	- Check for `JobStatus.DONE` and print error if any
	- Support qiskit-jku-provider
	- Minor typos and more error handling
	- Handle Ast2CircException
	- Protect import of IBMQ Provider with try/except
	- Now reports own version with --qisjob_version
	- If `-n` then `-v -v -v` verbosity will print unrolled circuit
	- Added `no_unknown=True` so nuqasm2 will except on unparseable

v3.0:
	- Now installable with `setup.py`
	- with `-n` switch, uses nuqasm2 to assemble qasm code
	- Command once installed is `qisjob`
	- Support for new Aer GPU simulator
	- Added switches:
		`--circuit_layout`
		`--statevector_gpu`
		`--showsched`
	- Fixed bug in returning results of multiple experiment job
	- More exception catching

v2.8:
	- Main is now `qisjob` ... `qasm_job` is gone.
	- Warn, don't fail, if QuantumInspire not installed.
	- Support for Rigetti Forest
	- IBMQJobFailureError now prints job.error_mssage() and exits 100
	- Regularized exit codes and explanation

v2.7.1:
   - Support datetime for backend properties
   - Fix bug in single-job for quantum circuit file input
   - Remove extraneous printout of env

v2.7:
   - Supports loading QuantumCircuit Python code as well as QASM.
   - Factor into QisJob object.
   - Output cleanups.
   - --job* commands work better with QI now

v2.6.2:
- --job_id now dumps the whole job
- Cleanup of job display

v2.6.1:
- Cleanup of output from -j flag.

v2.6:
- --jobs gets joblist from --backend
- --job_id gets status of specific job from --backend
- --job_result prints results of specific job from --backend

v2.5:
- more sample files
- --properties now does to_dict() before pretty printing
- -g --configuration prints backend configuration
- -p and -g take backend name from -b switch
- various bug fixes

v2.4:
- support qiskit-qcgpu-simulator
- support aer unitary_simulator
- check for counts before trying to print

v2.3:
- refactor result handling
- plotting state city
- plotting histogram
- choose specific Aer simulator
- --status to show status
- --backends to list backends


qasm_job.py
"""qasm_job.py
Load from qasm source and run job with reporting
Copyright 2019 Jack Woehr jwoehr@softwoehr.com PO Box 51, Golden, CO 80402-0051
BSD-3 license -- See LICENSE which you should have received with this code.
THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES."""

import argparse
import sys
from qis_job import QisJob

EXPLANATION = """qasm_job.py : Loads from one or more qasm source files or
from a file containing a Qiskit QuantumCircuit definition in Python and runs as
experiments with reporting in CSV form. Can graph results as histogram or
state-city plot. Also can give info on backend properties, qiskit version,
show circuit transpilation, etc. Can run as multiple jobs or all as one job.
Exits 0 on success, 1 on argument error, 100 on runtime error.
Copyright 2019 Jack Woehr jwoehr@softwoehr.com PO Box 51, Golden, CO 80402-0051.
BSD-3 license -- See LICENSE which you should have received with this code.
THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
"""
PARSER = argparse.ArgumentParser(description=EXPLANATION)
GROUP = PARSER.add_mutually_exclusive_group()
GROUPB = PARSER.add_mutually_exclusive_group()
GROUP.add_argument("-i", "--ibmq", action="store_true",
                   help="Use best genuine IBMQ processor (default)")
GROUP.add_argument("-s", "--sim", action="store_true",
                   help="Use IBMQ qasm simulator")
GROUP.add_argument("-a", "--aer", action="store_true",
                   help="""Use QISKit Aer simulator.
                   Default is Aer statevector simulator.
                   Use -a --qasm-simulator to get Aer qasm simulator.
                   Use -a --unitary-simulator to get Aer unitary simulator.""")
GROUP.add_argument("--qcgpu", action="store_true",
                   help="""Use qcgpu simulator.
                   Default is statevector simulator.
                   Use --qcgpu --qasm-simulator to get qcgpu qasm simulator.""")
GROUP.add_argument("-b", "--backend", action="store",
                   help="Use specified IBMQ backend")
GROUPB.add_argument("--qasm_simulator", action="store_true",
                    help="""With -a or --qcgpu use qasm simulator
                    instead of statevector simulator""")
GROUPB.add_argument("--unitary_simulator", action="store_true",
                    help="""With -a use unitary simulator
                    instead of statevector simulator""")
PARSER.add_argument("--api_provider", action="store",
                    help="""Backend remote api provider,
                    currently supported are [IBMQ | QI | Forest].
                    Default is IBMQ.""", default="IBMQ")
PARSER.add_argument("--qvm", action="store_true",
                    help="""Use Forest local qvm simulator described by
                    -b backend, generally one of qasm_simulator or
                    statevector_simulator. Use --qvm_as to instruct the
                    simulator to emulate a specific Rigetti QPU""")
PARSER.add_argument("--qvm_as", action="store_true",
                    help="""Use Forest local qvm simulator to emulate the
                    specific Rigetti QPU described by -b backend. Use --qvm
                    to run the Forest local qvm simulator described by
                    -b backend.""")
PARSER.add_argument("--backends", action="store_true",
                    help="Print list of backends to stdout and exit 0")
PARSER.add_argument("-1", "--one_job", action="store_true",
                    help="Run all experiments as one job")
PARSER.add_argument("-c", "--credits", type=int, action="store", default=3,
                    help="Max credits to expend on each job, default is 3")
PARSER.add_argument("-d", "--datetime", type=str, action="store",
                    help="""Datetime 'year,month,day[,hour,min,sec]'
                    for -p,--properties""")
PARSER.add_argument("-g", "--configuration", action="store_true",
                    help="""Print configuration for backend specified by -b
                    to stdout and exit 0""")
PARSER.add_argument("-j", "--job", action="store_true",
                    help="Print your job's dictionary")
PARSER.add_argument("--jobs", type=int, action="store",
                    help="Print JOBS jobs and status for -b backend and exit 0")
PARSER.add_argument("--job_id", type=str, action="store",
                    help="Print job number JOB_ID for -b backend and exit 0")
PARSER.add_argument("--job_result", type=str, action="store",
                    help=""""Print result of job number JOB_RESULT for
                    -b backend and exit 0""")
PARSER.add_argument("-m", "--memory", action="store_true",
                    help="Print individual results of multishot experiment")
PARSER.add_argument("-o", "--outfile", action="store",
                    help="Write appending CSV to outfile, default is stdout")
PARSER.add_argument("-p", "--properties", action="store_true",
                    help="""Print properties for backend specified by -b to
                    stdout and exit 0""")
PARSER.add_argument("-q", "--qubits", type=int, action="store", default=5,
                    help="Number of qubits for the experiment, default is 5")
PARSER.add_argument("--qiskit_version", action="store_true",
                    help="Print Qiskit version and exit 0")
PARSER.add_argument("-r", "--result", action="store_true",
                    help="Print job result")
PARSER.add_argument("-t", "--shots", type=int, action="store", default=1024,
                    help="Number of shots for the experiment, default 1024, max 8192")
PARSER.add_argument("-v", "--verbose", action="count", default=0,
                    help="Increase verbosity each -v up to 3")
PARSER.add_argument("-x", "--transpile", action="store_true",
                    help="""Print circuit transpiled for chosen backend to stdout
                    before running job""")
PARSER.add_argument("--histogram", action="store_true",
                    help="""Write image file of histogram of experiment results""")
PARSER.add_argument("--plot_state_city", type=int, action="store",
                    help="""Write image file of state city plot of statevector to
                    PLOT_STATE_CITY decimal points""")
PARSER.add_argument("--figure_basename", type=str, action="store",
                    default='figout',
                    help="""basename including path (if any) for figure output,
                    default='figout', backend name, figure type, and timestamp
                    will be appended""")
PARSER.add_argument("--qasm", action="store_true",
                    help="Print qasm file to stdout before running job")
PARSER.add_argument("--qc", action="store",
                    help="Indicate circuit name of python-coded QuantumCircuit")
PARSER.add_argument("--status", action="store_true",
                    help="""Print status of chosen --backend to stdout
                    (default all backends)
                    of --api_provider (default IBMQ)
                    and exit 0""")
PARSER.add_argument("--token", action="store",
                    help="Use this token")
PARSER.add_argument("--url", action="store",
                    help="Use this url")
PARSER.add_argument("filepath", nargs='*',
                    help="Filepath(s) to 0 or more .qasm files, default is stdin")

ARGS = PARSER.parse_args()

AER = ARGS.aer
API_PROVIDER = ARGS.api_provider.upper()
ARGS = PARSER.parse_args()
BACKEND_NAME = ARGS.backend
BACKENDS = ARGS.backends
CONFIGURATION = ARGS.configuration
DATETIME = ARGS.datetime
FIGURE_BASENAME = ARGS.figure_basename
FILEPATH = ARGS.filepath
HISTOGRAM = ARGS.histogram
JOB = ARGS.job
JOB_ID = ARGS.job_id
JOB_RESULT = ARGS.job_result
JOBS = ARGS.jobs
MAX_CREDITS = ARGS.credits
MEMORY = ARGS.memory
ONE_JOB = ARGS.one_job
OUTFILE = ARGS.outfile
PLOT_STATE_CITY = ARGS.plot_state_city
PROPERTIES = ARGS.properties
QASM = ARGS.qasm
QASM_SIMULATOR = ARGS.qasm_simulator
QC_NAME = ARGS.qc
QCGPU = ARGS.qcgpu
QISKIT_VERSION = ARGS.qiskit_version
QUBITS = ARGS.qubits
QVM = ARGS.qvm
QVM_AS = ARGS.qvm_as
RESULT = ARGS.result
SHOTS = ARGS.shots
SIM = ARGS.sim
STATUS = ARGS.status
TOKEN = ARGS.token
TRANSPILE = ARGS.transpile
UNITARY_SIMULATOR = ARGS.unitary_simulator
URL = ARGS.url
VERBOSE = ARGS.verbose

QJ = QisJob(filepaths=FILEPATH,
            provider_name=API_PROVIDER,
            backend_name=BACKEND_NAME,
            token=TOKEN, url=URL,
            num_qubits=QUBITS, shots=SHOTS, max_credits=MAX_CREDITS,
            outfile_path=OUTFILE, one_job=ONE_JOB, qasm=QASM,
            use_aer=AER, use_qasm_simulator=QASM_SIMULATOR, use_unitary_simulator=UNITARY_SIMULATOR,
            qcgpu=QCGPU, use_sim=SIM, qvm=QVM, qvm_as=QVM_AS,
            qc_name=QC_NAME, xpile=TRANSPILE,
            print_job=JOB, memory=MEMORY, show_result=RESULT,
            jobs_status=JOBS, job_id=JOB_ID, job_result=JOB_RESULT,
            show_backends=BACKENDS, show_configuration=CONFIGURATION, show_properties=PROPERTIES,
            show_statuses=STATUS, date_time=DATETIME,
            print_histogram=HISTOGRAM, print_state_city=PLOT_STATE_CITY,
            figure_basename=FIGURE_BASENAME,
            show_q_version=QISKIT_VERSION, verbose=VERBOSE)

QJ.do_it()

sys.exit(0)
